text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,446 @@
+├── s
+│   ├── src
+│   │   └── lib.rs
+│   └── target
+│       ├── doc
+│       │   ├── help.html
+│       │   ├── .lock
+│       │   ├── s
+│       │   │   ├── all.html
+│       │   │   ├── index.html
+│       │   │   ├── sidebar-items.js
+│       │   │   └── struct.S.html
+│       │   ├── search.desc
+│       │   │   └── s
+│       │   │       └── s-desc-0-.js
+│       │   ├── settings.html
+│       │   └── src
+│       │       └── s
+│       │           └── lib.rs.html
+│       ├── doc.parts
+│       │   └── s
+│       │       └── crate-info.json
+│       └── libs.rmeta
+└── t
+    ├── src
+    │   └── lib.rs
+    └── target
+        ├── doc
+        │   ├── help.html
+        │   ├── .lock
+        │   ├── search.desc
+        │   │   └── t
+        │   │       └── t-desc-0-.js
+        │   ├── settings.html
+        │   ├── src
+        │   │   └── t
+        │   │       └── lib.rs.html
+        │   └── t
+        │       ├── all.html
+        │       ├── index.html
+        │       ├── sidebar-items.js
+        │       └── trait.T.html
+        ├── doc.parts
+        │   └── t
+        │       └── crate-info.json
+        └── libt.rmeta
+</pre>
+
+</details>
+
+
+# Reference-level explanation
+
+The existing cross-crate information files, like `search-index.js`, all are lists of elements, rendered in an specified way (e.g. as a JavaScript file with a JSON array or an HTML index page containing an unordered list). The current rustdoc (in `write_shared`) pushes the current crate's version of the CCI into the one that is already found in `doc`, and renders a new version. The rest of the proposal uses the term **part** to refer to the pre-merged, pre-rendered element of the CCI. This proposal does not add any new CCI or change their contents (modulo sorting order, whitespace).
+
+## New file: `crate-info.json`
+
+The `crate-info.json` file contains the unmerged contents of a single crates' version of their corresponding CCI. This file is written if the flag `--write-info-json=path/to/crate-info.json` is provided.
+
+`crate-info.json` is a JSON object with several key-value pairs. Every value is a JSON array. Every element of the JSON array is a two-element array: a destination filename relative to the doc root, and the representation of the part. The representation of the part depends on the type of CCI that it describes.
+
+* key: `version`
+
+The value will be a string encoding of a version number. If rustdoc is provided a `--include-info-json` flag that points to a `crate-info.json` file with a version number that it cannot support, it will fail. Rustdoc only guarantees that it will accept `crate-info.json` files written by the same version of rustdoc. There is no forward or backward compatibility. A single number prefixed with capital V is chosen for convenience, as `crate-info.json` does not make distinctions between breaking and non-breaking changes. Rustdoc is the only explicitly supported consumer of `crate-info.json`.
```

Comment:
Thanks for writing the section. Now we just need the rustdoc team to agree on a versioning system (because this is the one we use in the JSON output and we're discussing about changing it, so better fix it before an RFC to use it is approved ^^').

Reply:
No but if we could at least have the version format to be definitive, that'd make our (rustdoc) work much simpler if we can rely on this key whatever the rest of the file looks like.

Reply:
I agree that the content of `crate-info.json` is opaque. I think that how we determine whether rustdoc is pointed to an incompatible `crate-info.json`, and how we report that error message, is an implementation detail. The version number, just like the content of the file, will not be visible or consumable by users of rustdoc.

Reply:
Hum... That's a fair point. Ok, then I'm ok with this way of doing it. Then please add this explanation in the RFC text and remove the part about what the version is about, we'll discuss it in the implementation then.

Reply:
Okay, I removed many of the specific details about the contents of crate-info.json, including the version key, and explained that they were unstable.